File: common/src/main/java/org/nzbone/common/util/QuerySet.java
package org.nzbone.common.util;

import rx.Observable;

import java.util.HashSet;

public class QuerySet<T> extends HashSet<T> {

    @Override
    public String toString() {
        return Observable.from(this)
                .map(Object::toString)
                .reduce((c,n) -> c + "|" + n)
                .toBlocking()
                .single();
    }
}

File: common/src/main/java/org/nzbone/util/QuerySet.kt
package org.nzbone.util

import java.util.HashSet

fun <T> querySetOf(vararg values: T): QuerySet<T> {
    return QuerySet<T>().apply {
        addAll(values)
    }
}

fun <T> Set<T>.toQuerySet(): QuerySet<T> {
    return QuerySet<T>().apply {
        addAll(this)
    }
}

class QuerySet<T> : HashSet<T>() {

    override fun toString(): String {
        return this.joinToString(separator = "|") { it.toString() }
    }
}
